src/constraint.rs

```diff
@@ -1563,3 +1581,209 @@ impl ConstraintValidator for ValidValuesConstraint {
+                            }
+                            '}' => {
+                                sb.push(ch);
+                                // at this point occurrences are completely specified
+                                complete = true;
+                                break;
+                            }
+                            _ => return invalid_schema_error(format!("invalid character {}", ch)),
+                        }
+                    }
+
+                    if !complete {
+                        return invalid_schema_error("range quantifier missing '}'");
+                    }
+                }
+                _ => {}
+            }
+            if sb.len() > initial_length {
+                if let Some(ch) = si.peek().cloned() {
+                    match ch {
+                        '?' => return invalid_schema_error(format!("invalid character {}", ch)),
+
+                        '+' => return invalid_schema_error(format!("invalid character {}", ch)),
+                        _ => {}
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl ConstraintValidator for RegexConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        let value = match value.as_str() {
+            Some(string_value) => {
+                let re = Regex::new(r"\r").unwrap();
+                let result = re.replace_all(string_value, "\n");
+                result.to_string()
+            }
+            _ => {
+                // return Violation if value is not a string
+                return Err(Violation::new(
+                    "regex",
+                    ViolationCode::TypeMismatched,
+                    &format!("expected a string but found {}", value.ion_type()),
+                ));
+            }
+        };
+
+        if !self.expression.is_match(value.as_str()) {
+            return Err(Violation::new(
+                "regex",
+                ViolationCode::RegexMismatched,
+                &format!("{} doesn't match regex {}", value, self.expression),
+            ));
+        }
+
+        Ok(())
+    }
+}
+
+impl TryFrom<IslRegexConstraint> for RegexConstraint {
+    type Error = IonSchemaError;
+
+    fn try_from(isl_regex: IslRegexConstraint) -> Result<Self, Self::Error> {
+        let pattern = RegexConstraint::convert_to_pattern(isl_regex.expression().to_owned())?;
+
+        let regex = RegexBuilder::new(pattern.as_str())
+            .case_insensitive(isl_regex.case_insensitive())
+            .multi_line(isl_regex.multi_line())
+            .build()
+            .map_err(|e| {
+                invalid_schema_error_raw(format!("Invalid regex {}", isl_regex.expression()))
+            })?;
+
+        Ok(RegexConstraint::new(regex))
+    }
+}
+
+impl PartialEq for RegexConstraint {
+    fn eq(&self, other: &Self) -> bool {
+        self.expression.as_str().eq(other.expression.as_str())
```

Comment:
I meant to add the `PartialEq` because regex crate doesn't have one defined so I can not derive it. Although that being said it can definitely be an `Eq` instead of `PartialEq` because its just comparing both string expressions and can a have a proper equality.

Also, no the string comparison doesn't include the flags but thanks for pointing it out I will change it in the implementation here.